tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp

```diff
@@ -206,23 +207,57 @@ struct InitializeElement {
   }
 };
 
+namespace detail {
+
+template <typename Preconditioner>
+struct init_preconditioner {
+  using type = typename Preconditioner::initialize_element;
+};
+template <>
+struct init_preconditioner<void> {
+  using type = tmpl::list<>;
+};
+
+template <typename Preconditioner>
+struct run_preconditioner {
+  using type =
+      tmpl::list<ComputeOperatorAction<typename Preconditioner::fields_tag>,
+                 typename Preconditioner::prepare_solve,
+                 ::Actions::RepeatUntil<
+                     LinearSolver::Tags::HasConverged<
+                         typename Preconditioner::options_group>,
+                     tmpl::list<typename Preconditioner::prepare_step,
+                                ComputeOperatorAction<
+                                    typename Preconditioner::operand_tag>,
+                                typename Preconditioner::perform_step>>>;
+};
+template <>
+struct run_preconditioner<void> {
+  using type = tmpl::list<>;
+};
+
+}  // namespace detail
+
 template <typename Metavariables>
 struct ElementArray {
   using chare_type = Parallel::Algorithms::Array;
   using array_index = int;
   using metavariables = Metavariables;
   using linear_solver = typename Metavariables::linear_solver;
+  using preconditioner = typename Metavariables::preconditioner;
+
   // In each step of the algorithm we must provide A(p). The linear solver then
   // takes care of updating x and p, as well as the internal variables r, its
   // magnitude and the iteration step number.
   /// [action_list]
   using phase_dependent_action_list = tmpl::list<
       Parallel::PhaseActions<
           typename Metavariables::Phase, Metavariables::Phase::Initialization,
-          tmpl::list<InitializeElement,
-                     typename linear_solver::initialize_element,
-                     ComputeOperatorAction<fields_tag>,
-                     Parallel::Actions::TerminatePhase>>,
+          tmpl::list<
+              InitializeElement, typename linear_solver::initialize_element,
+              ComputeOperatorAction<fields_tag>,
+              tmpl::type_from<detail::init_preconditioner<preconditioner>>,
```

Comment:
`typename blah::type`. `tmpl::type_from` should be used on `tmpl::type_` only. It doesn't really save any characters and makes the code harder to read.